### PR TITLE
playwright: 1.27.1 -> 1.31.1

### PR DESCRIPTION
**Pull request by mguentner**
###### Description of changes

Updates playwright from 1.27.1 to 1.31.0. Tested by running (typescript) e2e tests and `playwright cr https://nixos.org`

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wegank** [comment]
@ofborg build playwright playwright.passthru.tests

**Comment by NickCao** [comment]
```
error: hash mismatch in fixed-output derivation '/nix/store/i5d93s160jv7wy3sdg7pf2xf1jlr0a4b-playwright-browsers-base.drv':
         specified: sha256-clpDSEZHVT1Vo1fYwKxPc2w35X+RKTUiIrbq8x2Uepg=
            got:    sha256-0ZRC1UpPLEw59hbKHkNe3VnnQOdA6nx7Xp/aziLcuv4=
```

**Comment by mguentner** [comment]
Can someone on x86_64-darwin please support to figure out what is going on here?

**Comment by mguentner** [comment]
NixOS x86_64: the only browser that works for me is `chromium`. I get the same error with `firefox` as @Gerschtli
At least the derivation only seems to package `chromium`, `firefox` and `ffmpeg-linux`

**Comment by SuperSandro2000** [comment]
> also the applied patches.

We cannot realistically apply them because they are gigantic blobs and unmanageable unless we pin the exact browser version from upstream and then we can also download their blob.

**Comment by mguentner** [comment]
I agree. For the record, it might be possible to extract the information [from here](https://github.com/microsoft/playwright/blob/main/browser_patches/firefox/UPSTREAM_CONFIG.sh), but this is still no guarantee that the result will work as expected.

Accordingly we should remove firefox from the derivation and only enable chromium as this is the only browser that currently works.

Still - I have no idea what the situation on darwin is as the derivation looks completely different. @SuperSandro2000 have you tested your initial work on darwin?

**Comment by mguentner** [comment]
I have removed `firefox` support (on Linux, see discussion above) and support for `darwin` altogether in two commits.
If it turns out that playwright actually works on darwin with the upstream binaries, then I can easily delete the most recent commit [14ad29e](https://github.com/NixOS/nixpkgs/pull/214939/commits/14ad29ea096d71d79964acde7f1e41730ace5e28).

@SuperSandro2000 :point_up:

**Comment by yrd** [comment]
Sorry for taking so long to review this…
We've had this discussion before here: https://github.com/microsoft/playwright/issues/5501 It would be nice to get some feedback from the Playwright maintainers, maybe we can include a small subset of their patches to get it to work. As for Darwin, I'll try to get around to testing it again… As far as I remember, the Playwright-provided binaries (which you also get when running `playwright install`) should work as-is there. Since it depends on system stuff anyway, I would argue it isn't really that reproducible anyway when using the browser bundle on Darwin so it should be fine to just drop it.

**Comment by SuperSandro2000** [comment]
> maybe we can include a small subset of their patches to get it to work

Last time I checked they had one gigantic patch file, instead of multipler smaller ones we could more easily apply.

File: pkgs/development/python-modules/playwright/default.nix [hunk 1]
```diff
@@ -79,43 +74,18 @@ let
...
-
-  browsers-linux = { withFirefox ? true, withChromium ? true }: let
+  browsers-linux = {}: let
     fontconfig = makeFontsConf {
       fontDirectories = [];
     };
-  in runCommand ("playwright-browsers"
-    + lib.optionalString (withFirefox && !withChromium) "-firefox"
-    + lib.optionalString (!withFirefox && withChromium) "-chromium")
+  in runCommand ("playwright-browsers")
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/playwright/default.nix:81]:
```suggestion
  in runCommand "playwright-browsers"
```

**Review by SuperSandro2000 — COMMENTED** [review]
please rebase. I don't really have a strong opinion about the changes. Should be alright.

**Comment by mguentner** [comment]
I have rebased against the current master and did some changes along the way:
* playwright-python is now at the latest 1.31.x release, I did not want to advance it further as already a dependency changed again (greenlet 1.x -> 2.x)
* I have re-ordered the commits for convenience while testing under darwin, @yrd if you drop the latest commit, you will have mac support again to report back whether the binaries simply work :+1: 

Regarding the firefox situation: I think that _if_ nixpkgs wants to support playwright + firefox, it should try to package the upstream patched binary and make it run instead of producing a similar patched binary with the nixpkgs firefox derivation. The main reason is that playwright is a e2e testing framework and environments should be quite reproducible so that it makes sense. Replacing binaries changes the environment drastically (like it does not work at all and all your tests fail under NixOS :laughing:). Patching them to make them work under NixOS only will produce NixOS issues like dynamic linker, patchelf etc. but not fundamental issues like version mismatch (playwright assumes it's firefox x, but nixpkgs replaced it with y) or missing APIs.
After this PR, https://github.com/NixOS/nixpkgs/pull/197899 should be rebased and firefox support can be added again through the upstream binary. WDYT?

**Comment by wegank** [comment]
There was @hamann's #225708 (and a follow-up #225758) a few hours ago, which fixed the x86_64-darwin builds, so I guess the last commit could be dropped?

**Comment by mguentner** [comment]
@wegank Ah nice, I did not see that! `darwin` support is back.

**Comment by Gerschtli** [comment]
FYI: There is now 1.32.3 available :)

**Comment by fczuardi** [comment]
so, is it possible to use playwright test on a nodejs project in nixos?

I have tried a bunch of different combination of pkgs, like playwright, playwright-driver, playwright-test, set up the env vars PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1 PLAYWRIGHT_SKIP_VALIDATE_HOST_REQUIREMENTS=true PLAYWRIGHT_BROWSERS_PATH=$(nix build --print-out-paths nixpkgs#playwright-driver.browsers)

and typing `playwright test` on a js project with tests give me all errors with Error: Failed to launch the browser process! spawn /home/fcz/.cache/puppeteer/chrome/linux-1108766/chrome-linux/chrome ENOENT

**Comment by Gerschtli** [comment]
You need to make sure that the nixpkgs version and the one installed via npm are using the exact same version of playwright. Maybe your npm installed one is more recent.

**Pull request merged**